### PR TITLE
Switch chat thread when joining breakout room

### PR DESCRIPTION
**Pull request by mgamis-msft**
# What
Implement switching chat thread when joining breakout room and also switching back to the origin call's chat thread when leaving breakout room

# Why
Breakout rooms feature

# How Tested
Tested on local CallWithChat sample:
https://github.com/user-attachments/assets/8bf56f12-f71f-4b94-b4a0-98505328d0f4

# Process & policy checklist

- [ ] I have updated the project documentation to reflect my changes if necessary.
- [ ] I have read the [CONTRIBUTING](https://github.com/Azure/communication-ui-library/blob/main/CONTRIBUTING.md) documentation.

**Is this a breaking change?**

- [ ] This change causes current functionality to break.

### CODE REVIEW

File: packages/chat-component-bindings/src/messageThreadSelector.ts [hunk 1]
```diff
@@ -317,7 +317,7 @@ const hasValidParticipant = (chatMessage: ChatMessageWithStatus): boolean =>
 export const messageThreadSelectorWithThread: () => MessageThreadSelector = () =>
   createSelector(
     [getUserId, getChatMessages, getLatestReadTime, getIsLargeGroup, getReadReceipts, getParticipants],
-    (userId, chatMessages, latestReadTime, isLargeGroup, readReceipts, participants) => {
+    (userId, chatMessages, latestReadTime, isLargeGroup, readReceipts = [], participants) => {
```
Comment (JamesBurnside) [inline packages/chat-component-bindings/src/messageThreadSelector.ts:320]: We should fix getReadReceipts instead so it is fixed for every selector.

getReadReceipts return type should be `ChatMessageReadReceipt[] | undefined`

File: packages/react-composites/src/composites/CallWithChatComposite/CallWithChatComposite.tsx [hunk 2]
```diff
@@ -577,28 +581,52 @@ const CallWithChatScreen = (props: CallWithChatScreenProps): JSX.Element => {
     ]
   );
 
-  const onRenderChatContent = useCallback(
-    (): JSX.Element => (
+  const onRenderChatContent = useCallback((): JSX.Element => {
+    /* @conditional-compile-remove(breakout-rooms) */
+    return callWithChatAdapter.getState().chat?.threadId ? (
```
Comment (JamesBurnside) [inline packages/react-composites/src/composites/CallWithChatComposite/CallWithChatComposite.tsx:586]: instead of `getState`, use the onStateChange and store threadId in a useState. Reason being is this should re-render when the threadId changes and hence threadId should be in the dependency array of this callback

File: packages/react-composites/src/composites/CallWithChatComposite/adapter/AzureCommunicationCallWithChatAdapter.ts [hunk 3]
```diff
@@ -177,18 +192,85 @@ export class AzureCommunicationCallWithChatAdapter implements CallWithChatAdapte
...
+    /* @conditional-compile-remove(breakout-rooms) */
+    this.callAdapter.on('breakoutRoomsUpdated', async (eventData: BreakoutRoomsEventData) => {
+      if (!eventData.data) {
+        return;
+      }
+      if (eventData.type === 'join') {
+        await this.breakoutRoomJoined(eventData.data);
+      } else if (eventData.type === 'assignedBreakoutRooms') {
+        if (eventData.data.state === 'closed') {
+          this.returnFromBreakoutRoom();
```
Comment (JamesBurnside) [inline packages/react-composites/src/composites/CallWithChatComposite/adapter/AzureCommunicationCallWithChatAdapter.ts:204]: nit: await this (not actually relevant as nothing else executes after this line, but if someone adds to this line it could be easily missed so better to add it)

```suggestion
          await this.returnFromBreakoutRoom();
```

File: packages/react-composites/src/composites/CallWithChatComposite/adapter/AzureCommunicationCallWithChatAdapter.ts [hunk 4]
```diff
@@ -177,18 +192,85 @@ export class AzureCommunicationCallWithChatAdapter implements CallWithChatAdapte
     };
 
     this.callAdapter.onStateChange(onCallStateChange);
+    /* @conditional-compile-remove(breakout-rooms) */
+    this.callAdapter.on('breakoutRoomsUpdated', async (eventData: BreakoutRoomsEventData) => {
+      if (!eventData.data) {
```
Comment (JamesBurnside) [inline packages/react-composites/src/composites/CallWithChatComposite/adapter/AzureCommunicationCallWithChatAdapter.ts:197]: Out of interest, when is there not data?

File: packages/react-composites/src/composites/common/utils.tsx [hunk 5]
```diff
@@ -60,3 +60,18 @@ export const useMinMaxDragPosition = (modalLayerHostId: string, rtl?: boolean):
...
+/* @conditional-compile-remove(breakout-rooms) */
+/**
+ * @private
+ */
+export const busyWait = async (checkCondition: () => boolean, retryLimit?: number): Promise<void> => {
+  const delayMs = 500;
+  let retryCount = 0;
+  while (!checkCondition()) {
+    await new Promise((resolve) => setTimeout(resolve, delayMs));
+    if (retryLimit && retryCount++ >= retryLimit) {
```
Comment (JamesBurnside) [inline packages/react-composites/src/composites/common/utils.tsx:73]: nit: if retryLimit is zero this will fail

File: packages/react-composites/src/composites/CallWithChatComposite/adapter/AzureCommunicationCallWithChatAdapter.ts [hunk 6]
```diff
@@ -177,18 +192,85 @@ export class AzureCommunicationCallWithChatAdapter implements CallWithChatAdapte
...
+      }
+    });
     this.onCallStateChange = onCallStateChange;
   }
 
+  /* @conditional-compile-remove(breakout-rooms) */
+  private async breakoutRoomJoined(call: Call | TeamsCall): Promise<void> {
+    const targetThreadId = call.info.threadId;
+    // If the chat adapter is not on the target thread then we need to switch to the breakout room chat adapter
+    if (targetThreadId && this.chatAdapter && this.chatAdapter.getState().thread.threadId !== targetThreadId) {
```
Comment (JamesBurnside) [inline packages/react-composites/src/composites/CallWithChatComposite/adapter/AzureCommunicationCallWithChatAdapter.ts:215]: (reviewer todo: adding comment to remind me that I want to review if there's any race conditions in the existing logic that waits for threadId to come in from the call object)

File: packages/react-composites/src/composites/CallWithChatComposite/CallWithChatComposite.tsx (hunk 2 above)
Comment (mgamis-msft) [reply packages/react-composites/src/composites/CallWithChatComposite/CallWithChatComposite.tsx:586]: I also think getState isn't reliable for to get timely re-renders. So I tried to store the threadId in a useState like you said:
image

But for some reason it did not set the threadId to undefined (no spinner) like I expected as seen in this test run:
https://github.com/user-attachments/assets/c980f330-878e-4dfc-9601-45882fb061ea.

I will try other ways to not use `getState`

File: packages/react-composites/src/composites/CallWithChatComposite/adapter/AzureCommunicationCallWithChatAdapter.ts (hunk 4 above)
Comment (mgamis-msft) [reply packages/react-composites/src/composites/CallWithChatComposite/adapter/AzureCommunicationCallWithChatAdapter.ts:197]: I've never encountered a case where data was undefined. I will ask Sravanthi.

File: packages/react-composites/src/composites/common/utils.tsx (hunk 5 above)
Comment (mgamis-msft) [reply packages/react-composites/src/composites/common/utils.tsx:73]: It should break when it is `retryLimit` is 0 because `retryCount` would be >= than 0 on the first iteration. If you mean when retryLimit is undefined which would be the default then yes it would run forever. Should I prevent that and set a default retryLimit?

File: packages/chat-component-bindings/src/messageThreadSelector.ts (hunk 1 above)
Comment (mgamis-msft) [reply packages/chat-component-bindings/src/messageThreadSelector.ts:320]: That makes sense. Changed return type of getReadReceipts to `ChatMessageReadReceipt[] | undefined`.

File: packages/react-composites/src/composites/common/utils.tsx (hunk 5 above)
Comment (JamesBurnside) [reply packages/react-composites/src/composites/common/utils.tsx:76]: > It should break when it is retryLimit is 0 because retryCount would be >= than 0

This won't happen, when it's 0 it will evaulate as
```
if (0 && retryCount++ >= 0)
```

So the first part (`0 && ..`) will always evaluate to false as `if (0)` is false

Comment (JamesBurnside) [reply packages/react-composites/src/composites/common/utils.tsx:76]: So something like this would fix:
```suggestion
    if (!retryLimit || (retryLimit && retryCount++ >= retryLimit)) {
```

File: packages/react-composites/src/composites/CallWithChatComposite/CallWithChatComposite.tsx (hunk 2 above)
Comment (JamesBurnside) [reply packages/react-composites/src/composites/CallWithChatComposite/CallWithChatComposite.tsx:586]: Hmmm, your screenshot should've worked...

- Is callwithchatadpater.onStateChange not firing when the chat changes?
- Was threadId being set but for some reason the onRenderChatContent wasn't firing?

Comment (mgamis-msft) [reply packages/react-composites/src/composites/CallWithChatComposite/CallWithChatComposite.tsx]: Ah ok. I had to first unsub to main meeting chat adapter state changes to get it to work
image
The main meeting chat adapter would just update the callWithChatAdapterState right after I unset the chat state.
Also I just used the existing useState `isChatInitialized` to decide whether to show the spinner.
Looks much better now. Thanks!

File: packages/react-composites/src/composites/common/utils.tsx (hunk 5 above)
Comment (mgamis-msft) [reply packages/react-composites/src/composites/common/utils.tsx]: I see. Forgot 0 was falsy. Also covers NaN which is falsy. Updated.

File: packages/react-composites/src/composites/CallComposite/adapter/AzureCommunicationCallAdapter.ts [hunk 7]
```diff
@@ -1352,14 +1365,41 @@ export class AzureCommunicationCallAdapter<AgentType extends CallAgent | TeamsCa
 
   /* @conditional-compile-remove(breakout-rooms) */
   private assignedBreakoutRoomUpdated(breakoutRoom: BreakoutRoom): void {
+    if (this.call?.id === undefined) {
+      return;
+    }
```
Comment (mgamis-msft) [inline packages/react-composites/src/composites/CallComposite/adapter/AzureCommunicationCallAdapter.ts:1370]: Look into why need this.

File: packages/react-composites/src/composites/CallComposite/adapter/AzureCommunicationCallAdapter.ts [hunk 8]
```diff
@@ -1352,14 +1365,41 @@ export class AzureCommunicationCallAdapter<AgentType extends CallAgent | TeamsCa
...
-    this.processNewCall(call);
+    if (this.call?.id !== call.id) {
+      this.processNewCall(call);
+    }
+    // Hang up other breakout room calls in case we are joining a new breakout room while already in one
+    this.hangupOtherBreakoutRoomCalls(call);
+  }
+
+  /* @conditional-compile-remove(breakout-rooms) */
+  private hangupOtherBreakoutRoomCalls(breakoutRoomCall: Call | TeamsCall): void {
```
Comment (mgamis-msft) [inline packages/react-composites/src/composites/CallComposite/adapter/AzureCommunicationCallAdapter.ts:1386]: rename breakoutRoomCall to -> currentBreakoutRoomCall

**Review by dmceachernmsft — COMMENTED** [review]
looks good to me! just waiting to approve on the comment you added in our call

File: packages/react-composites/src/composites/CallWithChatComposite/CallWithChatComposite.tsx [hunk 9]
```diff
@@ -577,28 +581,47 @@ const CallWithChatScreen = (props: CallWithChatScreenProps): JSX.Element => {
...
   );
 
-  const onRenderChatContent = useCallback(
-    (): JSX.Element => (
+  const onRenderChatContent = useCallback((): JSX.Element => {
+    /* @conditional-compile-remove(breakout-rooms) */
+    if (!isChatInitialized) {
+      return (
+        <Stack styles={chatSpinnerContainerStyles}>
+          <Spinner size={SpinnerSize.large} />
```
Comment (JamesBurnside) [inline packages/react-composites/src/composites/CallWithChatComposite/CallWithChatComposite.tsx:589]: Any a11y requirements to state why a spinner is showing similar to an alt text for image?

File: packages/react-composites/src/composites/CallComposite/adapter/AzureCommunicationCallAdapter.ts (hunk 8 above)
Comment (JamesBurnside) [reply packages/react-composites/src/composites/CallComposite/adapter/AzureCommunicationCallAdapter.ts:1386]: nit: just pass `currentBreakoutRoomCallId` if just the id is used

File: packages/react-composites/src/composites/CallComposite/adapter/AzureCommunicationCallAdapter.ts (hunk 7 above)
Comment (mgamis-msft) [reply packages/react-composites/src/composites/CallComposite/adapter/AzureCommunicationCallAdapter.ts]: Tested the bug bash scenarios after removing this and everything passes.

This condition might have actually caused some test cases to fail in the initial bug bash when a breakout rooms call ends (when closed or timed out) and the tester did not come back to the meeting.

File: packages/react-composites/src/composites/CallComposite/adapter/AzureCommunicationCallAdapter.ts [hunk 10]
```diff
@@ -1060,18 +1060,31 @@ export class AzureCommunicationCallAdapter<AgentType extends CallAgent | TeamsCa
...
-    // Find main meeting call from call agent from the this call state
-    const mainMeetingCall = this.callAgent?.calls.find((callAgentCall) => {
-      return callAgentCall.id === callState?.breakoutRooms?.breakoutRoomOriginCallId;
+    // Current call state of breakout room may not be present in calls array if the breakout room call is ended.
+    // So let's try to find the call state from callsEnded array
+    if (callState === undefined) {
+      callState = this.call
+        ? Object.values(this.callClient.getState().callsEnded).find((call) => call.id === this.call?.id)
+        : undefined;
+    }
```
[the remark this answers is not in the document]
Comment (mgamis-msft) [reply packages/react-composites/src/composites/CallComposite/adapter/AzureCommunicationCallAdapter.ts]: Your refactor looks cleaner. Updated.

File: packages/react-composites/src/composites/CallWithChatComposite/adapter/AzureCommunicationCallWithChatAdapter.ts [hunk 11]
```diff
@@ -177,18 +192,115 @@ export class AzureCommunicationCallWithChatAdapter implements CallWithChatAdapte
...
+      // is also seen in Teams
+      // Check up to 20 times every 500ms and then continue.
+      await busyWait(() => {
+        // If the call adapter's call id has been changed, then stop waiting and don't update the chat adapter
+        if (this.context.getState().call?.id !== call.id) {
+          return true;
+        }
+
+        const userAddedToThread =
+          !!this.breakoutRoomChatAdapter?.getState().thread.participants[
```
Comment (JamesBurnside) [inline packages/react-composites/src/composites/CallWithChatComposite/adapter/AzureCommunicationCallWithChatAdapter.ts:243]: So can we access the `.thread.participants` without having been added to the thread yet?

Comment (JamesBurnside) [reply packages/react-composites/src/composites/CallWithChatComposite/adapter/AzureCommunicationCallWithChatAdapter.ts:243]: I suppose its just empty array?

Comment (mgamis-msft) [reply packages/react-composites/src/composites/CallWithChatComposite/adapter/AzureCommunicationCallWithChatAdapter.ts:243]: Yes it is just an empty array

File: packages/react-composites/src/composites/CallComposite/adapter/AzureCommunicationCallAdapter.ts [hunk 12]
```diff
@@ -1060,18 +1060,31 @@ export class AzureCommunicationCallAdapter<AgentType extends CallAgent | TeamsCa
...
+    const originCallId = callState?.breakoutRooms?.breakoutRoomOriginCallId;
+
+    // Find origin call from call agent
+    const originCall = this.callAgent?.calls.find((callAgentCall) => {
+      return callAgentCall.id === originCallId;
     });
-    // If a main meeting call exists then process that call and resume
-    if (mainMeetingCall) {
+
+    // If an origin call exists then process that call and resume
```
[the remark this answers is not in the document]
Comment (mgamis-msft) [reply packages/react-composites/src/composites/CallComposite/adapter/AzureCommunicationCallAdapter.ts:1078]: I'll throw an error since a call state of a breakout room should always have a origin call.

File: packages/react-composites/src/composites/CallWithChatComposite/adapter/AzureCommunicationCallWithChatAdapter.ts [hunk 13]
```diff
@@ -1044,6 +1161,17 @@ export const createAzureCommunicationCallWithChatAdapter = async ({
       'CallWithChat' as _TelemetryImplementationHint
     );
     callWithChatAdapter.setChatAdapterPromise(chatAdapterPromise);
+    /* @conditional-compile-remove(breakout-rooms) */
+    callWithChatAdapter.setCreateChatAdapterCallback((threadId: string) =>
```
Comment (JamesBurnside) [inline packages/react-composites/src/composites/CallWithChatComposite/adapter/AzureCommunicationCallWithChatAdapter.ts:1165]: Should thie be inside `if (chatThreadAdapter.isCallInfoRequired())` or should it be after the else? (Does this not always need set?)

File: packages/react-composites/src/composites/CallWithChatComposite/CallWithChatComposite.tsx (hunk 9 above)
Comment (dmceachernmsft) [reply packages/react-composites/src/composites/CallWithChatComposite/CallWithChatComposite.tsx:589]: For this spinner is there any label information to add like "Fetching Chat Thread" or something?

File: packages/react-composites/src/composites/CallWithChatComposite/adapter/AzureCommunicationCallWithChatAdapter.ts (hunk 13 above)
Comment (mgamis-msft) [reply packages/react-composites/src/composites/CallWithChatComposite/adapter/AzureCommunicationCallWithChatAdapter.ts:1167]: I suppose it should always be set even though breakout rooms for now is only for Teams meetings. Updated.

**Review by JamesBurnside — APPROVED** [review]
It's a tad hard to follow with the dynamic chat threads, create promises and the way threadIds can come in later than the calls, but it all looks good!
If possible, it would be good to try simplify the various chat thread extractions and swapping.
I guess what would be ideal to see is more defined boundaries where we have:
- may start with a threadId
- call may dynamically change in call, for any call threadIds can change at any time
  - on anytime call changes, threadId is changed + subscription for threadId changes.

```
function onThreadIdChanged() {
   ... set subcriptions etc. and push to the composite there has been chat updates
}

function createCallWithChatAdapter(threadId?: string) {
   ...
   if (threadId) => setChatAdapter(threadId)
   ...

  subscribeToThreadIdChange();
  subscribeToCallEvents();
}

function updateCall() {
  ...
  if (call.threadId) => onThreadIdChanged(threadId)
  subscribeToCallEvents(); // including call.threadId changing
}

function switchToBreakoutRoom() {
  this.updateThreadId(unset)
  await joinBreakoutRoom.then => this.updateThreadId(breakoutRoom.threadId)
}

function subscribeToCallEvents() {
  call.onThreadChanged => onThreadIdChanged();
}
```

But that could be a large refactor so can leave out of this PR

**Comment by mgamis-msft** [comment]
> It's a tad hard to follow with the dynamic chat threads, create promises and the way threadIds can come in later than the calls, but it all looks good! If possible, it would be good to try simplify the various chat thread extractions and swapping. I guess what would be ideal to see is more defined boundaries where we have:
> 
> * starting adapter may be created with a threadId
> * call may dynamically change in call, for any call threadIds can change at any time
>   
>   * on anytime call changes, threadId is changed + subscription for threadId changes.
> * Anytime threadId changes, we update chatAdapter and push the state
> 
> ```
> function onThreadIdChanged() {
>    ... set subcriptions etc. and push to the composite there has been chat updates
> }
> 
> function createCallWithChatAdapter(threadId?: string) {
>    ...
>    if (threadId) => setChatAdapter(threadId)
>    ...
> 
>   subscribeToThreadIdChange();
>   subscribeToCallEvents();
> }
> 
> function updateCall() {
>   ...
>   if (call.threadId) => onThreadIdChanged(threadId)
>   subscribeToCallEvents(); // including call.threadId changing
> }
> 
> function switchToBreakoutRoom() {
>   this.updateThreadId(unset)
>   await joinBreakoutRoom.then => this.updateThreadId(breakoutRoom.threadId)
> }
> 
> function subscribeToCallEvents() {
>   call.onThreadChanged => onThreadIdChanged();
> }
> ```
> 
> But that could be a large refactor so can leave out of this PR

That's totally fair. Some of the functions are handling to many scenarios and are too big. Created an [ADO task](https://skype.visualstudio.com/SPOOL/_workitems/edit/3855271) for me to later refactor based on this comment.

File: packages/react-composites/src/composites/CallWithChatComposite/CallWithChatComposite.tsx (hunk 9 above)
Comment (mgamis-msft) [reply packages/react-composites/src/composites/CallWithChatComposite/CallWithChatComposite.tsx:589]: I will add the label Teams has
image
and this is what it looks like in our composite
image

**Review by dmceachernmsft — APPROVED** [review]
